lib/RestructureCFG/InlineDivergentScopesPass.cpp

```diff
@@ -0,0 +1,390 @@
+          Switch->setDefaultDest(CaseIt->getCaseSuccessor());
+          Switch->removeCase(CaseIt);
+          break;
+        }
+      }
+    }
+
+    // Handle the simplification when `PlaceHolderTarget` is part the standard
+    // `case`s
+    for (auto CaseIt = Switch->case_begin(); CaseIt != Switch->case_end();) {
+      if (CaseIt->getCaseSuccessor() == PlaceHolderTarget) {
+
+        // We do not want to have a situation where the `PlaceHolderTarget` is
+        // both the `default` successor of a `switch` and one of its standard
+        // case
+        CaseIt = Switch->removeCase(CaseIt);
+      } else {
+        ++CaseIt;
+      }
+    }
+
+    // It should never be the case that we end up with a `switch` having only
+    // `PlaceHolderTarget` as its successor
+    if (Switch->getNumCases() == 0
+        and Switch->getDefaultDest() == PlaceHolderTarget) {
+      revng_abort();
+    }
+  }
+}
+
+/// We define a `Descriptor` for the information needed to identify a
+/// `Divergence`. Specifically, the divergence is composed by a set of divergent
+/// edges, and by the exit node which is divergent wrt. those divergent exits.
+using DivergentEdgesDescriptor = std::pair<BasicBlock *,
+                                           SmallSet<BasicBlock *, 4>>;
+using DivergenceDescriptorT = std::pair<DivergentEdgesDescriptor, BasicBlock *>;
+
+/// Helper function that tries to identify a `DivergenceDescriptor`
+static std::optional<DivergenceDescriptorT>
+electDivergence(BasicBlock *Candidate,
+                BasicBlock *Exit,
+                DomTreeOnView<BasicBlock, Scope> &DomTree) {
+
+  // We employ this `std::optional` as return value, in order to signal if we
+  // identify a candidate for IDS here
+  std::optional<DivergenceDescriptorT> DivergenceDescriptor;
+
+  // Check if it is the conditional making it a divergent node
+  if (isConditional(Candidate) and DomTree.dominates(Candidate, Exit)) {
+    using ScopeBlock = Scope<BasicBlock *>;
+
+    SetVector<BasicBlock *> Successors;
+    for (BasicBlock *Successor : children<ScopeBlock>(Candidate)) {
+      Successors.insert(Successor);
+    }
+
+    std::map<BasicBlock *, SmallSet<BasicBlock *, 4>> ReachablesFromSuccessor;
+    size_t SuccessorsSize = std::distance(Successors.begin(), Successors.end());
+
+    for (auto *Successor : Successors) {
+      SmallSet<BasicBlock *, 4> ReachableExits;
+
+      // Explore all the exits that are reachable from each Successor.
+      // For having a divergent exit, we need to find a set of
+      // successors that reach only the exit under analysis
+      for (auto *DFSNode : depth_first(Scope<BasicBlock *>(Successor))) {
+        if (isExit(DFSNode)) {
+          ReachableExits.insert(DFSNode);
+        }
+      }
+
+      ReachablesFromSuccessor[Successor] = ReachableExits;
+    }
+
+    SmallSet<BasicBlock *, 4> DivergentSuccessors;
+    for (const auto &[Successor, ReachableExits] : ReachablesFromSuccessor) {
```

Comment:
You're iterating on a `std::map` whose key is a `BasicBlock *`. This introduces non-determinism.
Turn the map into a vector or something.

Reply:
Ok makes sense.